timesketch/ui/graphs/cytoscape.component.ts

```diff
@@ -0,0 +1,224 @@
+  @Output() touchend = new EventEmitter<Evt['touchend']>()
+
+  // higher-level user input device events
+  @Output() tapstart = new EventEmitter<Evt['mousedown' | 'touchstart']>()
+  @Output() vmousedown = new EventEmitter<Evt['mousedown' | 'touchstart']>()
+  @Output() tapdrag = new EventEmitter<Evt['touchmove' | 'mousemove']>()
+  @Output() vmousemove = new EventEmitter<Evt['touchmove' | 'mousemove']>()
+  @Output() tapdragover = new EventEmitter<Evt['touchmove' | 'mousemove' | 'mouseover']>()
+  @Output() tapdragout = new EventEmitter<Evt['touchmove' | 'mousemove' | 'mouseout']>()
+  @Output() tapend = new EventEmitter<Evt['mouseup' | 'touchend']>()
+  @Output() vmouseup = new EventEmitter<Evt['mouseup' | 'touchend']>()
+  @Output() tap = new EventEmitter<Evt['click' | 'touchend']>()
+  @Output() vclick = new EventEmitter<Evt['click' | 'touchend']>()
+  @Output() taphold = new EventEmitter<Evt['taphold']>()
+  @Output() cxttapstart = new EventEmitter<Evt['mousedown' | 'touchstart']>()
+  @Output() cxttapend = new EventEmitter<Evt['mouseup' | 'touchend']>()
+  @Output() cxttap = new EventEmitter<Evt['click' | 'touchend']>()
+  @Output() cxtdrag = new EventEmitter<Evt['touchmove' | 'mousemove']>()
+  @Output() cxtdragover = new EventEmitter<Evt['touchmove' | 'mousemove' | 'mouseover']>()
+  @Output() cxtdragout = new EventEmitter<Evt['touchmove' | 'mousemove' | 'mouseout']>()
+  @Output() boxstart = new EventEmitter<Evt['boxstart']>()
+  @Output() boxend = new EventEmitter<Evt['boxend']>()
+  @Output() boxselect = new EventEmitter<Evt['boxselect']>()
+  @Output() box = new EventEmitter<Evt['box']>()
+
+  // graph events
+  @Output() layoutstart = new EventEmitter<Evt['layoutstart']>()
+  @Output() layoutready = new EventEmitter<Evt['layoutready']>()
+  @Output() layoutstop = new EventEmitter<Evt['layoutstop']>()
+  @Output() ready = new EventEmitter<Evt['ready']>()
+  @Output() destroy = new EventEmitter<Evt['destroy']>()
+  @Output() render = new EventEmitter<Evt['render']>()
+  @Output() resize = new EventEmitter<Evt['resize']>()
+
+  // tslint:enable:no-unused-variable
+
+  constructor(el: ElementRef) {
+    this.element = el.nativeElement
+  }
+
+  initCytoscape() {
+    const all_options = [].concat(immutable_options, mutable_options)
+    const options: Opt = {}
+    if (this.element) {
+      options['container'] = this.element.childNodes[0]
+    }
+    for (const k of all_options) {
+      options[k] = this[k]
+    }
+    this._cy = cytoscape(options)
+    this.initEvents()
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    const changed = (k) =>
+      changes[k].previousValue !== changes[k].currentValue
+    const options: Opt = {}
+    let needs_rebuild = false
+    for (const k of Object.keys(changes)) {
+      if (changed(k)) {
+        options[k] = changes[k].currentValue
+      }
+      if (immutable_options.indexOf(k) > -1 && this.cy) {
+        needs_rebuild = true
+        break
+      }
+    }
+    if (this.cy) {
```

Comment:
If this.cy doesn't exists, we don't need to do any of the things before, right? So we could move the if up to be in the first line and exit early, e.g.

if (!this.cy) return;  // We need an existing instance to apply changes to.

Reply:
Done.